File: Leetcode Easy/53MaximumSubarray.py
#https://leetcode.com/problems/maximum-subarray/description/
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        curSum, maxSum = 0, nums[0]
        for num in nums:
            curSum = max(curSum, 0)
            curSum += num
            maxSum = max(curSum, maxSum)
        return maxSum

File: C++/Leetcode Medium/53MaximumSubArray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int res = nums[0];
        int curr = 0;
        int l = 0; 
        for (int r = 0; r < nums.size(); ++r) {
            while (curr < 0 && l < r) curr -= nums[l++];
            curr += nums[r];
            res = max(res, curr);
        }
        return res;
    }
};